File: c/dictionary/dictionary.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "dictionary.h"

void expand(Dict *dict);
unsigned long hash(char s[]);

void Dict_init(Dict *dict) {
    dict->size = DICT_SIZE;

    // initialize array to be flagged empty
    for(int i = 0; i < dict->size; i++) {
        dict->contains[i] = 0;
    }
}

void Dict_set(Dict *dict, char *key, int value) {
    int index = hash(key) % dict->size;

    if(dict->contains[index] == 0) {
        Pair *p = malloc(sizeof(Pair));
        p->key = strdup(key);
        p->value = value;
        p->next = NULL;

        // inserting a new value into an empty spot
        dict->data[index] = *p;

        // flag the bucket we just added to
        dict->contains[index] = 1;
    } else if(strcmp(dict->data[index].key, key) == 0) {
        // update an existing value
        dict->data[index].value = value;
    } else {
        Pair *last = &dict->data[index];

        while(last->next != NULL) {
            last = last->next;
        }

        // append it to the end of the list
        last->next = malloc(sizeof(Pair));
        last->next->key = strdup(key);
        last->next->value = value;
        last->next->next = NULL;
    }
}

int Dict_get(Dict *dict, char *key) {
    int index = hash(key) % dict->size;

    if(dict->contains[index] == 0) {
        // we don't have it
        printf("Attempt to access non-existing key %s in Dict (non-collision)\n", key);
        return 0;
    } else  {
        Pair* curr = &dict->data[index];

        // is it in the first node of this bucket?
        if(strcmp(curr->key, key) == 0) {
            return curr->value;
        }

        // collision -- check the other nodes, if any
        while(curr != NULL) {
            if(strcmp(curr->key, key) == 0) {
                return curr->value;
            } else {
                curr = curr->next;
            }
        }

        // couldn't find it
        printf("Attempt to access non-existing key %s in Dict (collision)\n", key);
        return 0;
    }
}

unsigned long hash(char str[])
{
    unsigned long hash = 5381;

    for(int i = 0; i < strlen(str); i++) {
        int c = str[i];

        hash = ((hash << 5) + hash) ^ c; /* hash * 33 ^ c */
    }

    return hash;
}

File: haskell/src/Dictionary/Dictionary.hs
module Dictionary.Dictionary where

import Data.Maybe

import qualified Core.LinkedList as LinkedList
import Core.Helper
import Dictionary.StringHash

data Dictionary v = Dictionary{ buckets :: [LinkedList.List (String, v)], size :: Int }

newDict :: Dictionary a
newDict = (Dictionary (arrayOf LinkedList.EmptyNode defaultSize) 0)
  where
    arrayOf :: a -> Int -> [a]
    arrayOf _ 0 = []
    arrayOf a i = a : arrayOf a (i - 1)

    defaultSize = 1024

set :: Dictionary a -> String -> a -> Dictionary a
set d k v = let updatedDict = setHelper d k v
            in
              -- after updating the dictionary, make sure we
              -- don't need to expand it (using saturation limit of 70%)
              if isSaturated updatedDict
              then updatedDict
              else updatedDict
  where
    bucketCount :: Dictionary a -> Int
    bucketCount (Dictionary b _) = length b

    saturationLimit :: Dictionary a -> Int
    saturationLimit d' = bucketCount d'

    isSaturated :: Dictionary a -> Bool
    isSaturated d'@(Dictionary _ s) = s > saturationLimit d'

    isBucketEmpty :: LinkedList.List (String, v) -> Bool
    isBucketEmpty (LinkedList.Node _ _) = True
    isBucketEmpty LinkedList.EmptyNode   = False

    matchingKey :: String -> (String, v) -> Bool
    matchingKey s p = s == fst p

    -- returns a new dictionary equivalent to the given one,
    -- except it has twice as many buckets
    expand :: Dictionary a -> Dictionary a
    expand d'@(Dictionary b' _) = let resized = (Dictionary (arrayOf LinkedList.EmptyNode $ 2 * length b') 0) :: Dictionary a
                                  in copy d' resized -- copy the existing dict into a new dict
                                                     -- with twice as many buckets
      where
        arrayOf :: a -> Int -> [a]
        arrayOf _ 0 = []
        arrayOf a i = a : arrayOf a (i - 1)

    setHelper :: Dictionary a -> String -> a -> Dictionary a
    setHelper (Dictionary b' s') k' v'
      | isBucketEmpty bucket  = (Dictionary (replaceElementAt b' index newNode) (s' + 1)) -- no collision
      | isJust keyMatchedNode = (Dictionary (replaceElementAt b' index $ updateValue bucket (k', v')) s') -- since it's an update, no size change
      | otherwise             = (Dictionary (replaceElementAt b' index $ LinkedList.append bucket (k', v')) (s' + 1)) -- add this node to the bucket and update size
        where
          index = hash k' `mod` length b'
          bucket = b' !! index

          keyMatchedNode = LinkedList.findNode (matchingKey k') bucket
          newNode = LinkedList.newNode (k', v')

          updateValue :: (LinkedList.List (String, a)) -> (String, a) -> (LinkedList.List (String, a))
          updateValue l p@(k'', _) = LinkedList.replaceWhere (matchingKey k'') p l

-- copy the contents of d into d'
copy :: Dictionary a -> Dictionary a -> Dictionary a
copy (Dictionary b _) d = foldl (\curr next -> copyListIntoDict curr next) d b
  where
    copyListIntoDict :: Dictionary a -> LinkedList.List (String, a) -> Dictionary a
    copyListIntoDict d' LinkedList.EmptyNode         = d'
    copyListIntoDict d' (LinkedList.Node (k, v) n) = set (copyListIntoDict d' n) k v

get :: Dictionary a -> String -> Maybe a
get (Dictionary b _) k
  | isJust keyMatchedNode = Just $ snd (fromJust keyMatchedNode)
  | otherwise = Nothing
  where
    index = hash k `mod` length b
    bucket = b !! index

    matchingKey :: String -> (String, v) -> Bool
    matchingKey s p = s == fst p

    keyMatchedNode = LinkedList.find (matchingKey k) bucket

containsKey :: Dictionary a -> String -> Bool
containsKey (Dictionary b _) k = isJust keyMatchedNode
  where
    index = hash k `mod` length b
    bucket = b !! index

    matchingKey :: String -> (String, v) -> Bool
    matchingKey s p = s == fst p

    keyMatchedNode = LinkedList.find (matchingKey k) bucket

keys :: Dictionary a -> [String]
keys (Dictionary b _) = foldl (\curr next -> curr ++ bucketKeys next) [] b
  where
    bucketKeys :: (LinkedList.List (String, a) -> [String])
    bucketKeys = (map fst) . LinkedList.toList

pairs :: Dictionary a -> [(String, a)]
pairs (Dictionary b _) = foldl (\curr next -> curr ++ LinkedList.toList next) [] b

toList :: Dictionary a -> [(String, a)]
toList (Dictionary b _) = foldl (++) [] (map (LinkedList.toList) b)
